crates/node/src/networking/download_manager.rs

```diff
@@ -8,12 +9,101 @@ use hyper::server::conn::http1;
 use hyper::service::service_fn;
 use hyper::{Request, Response, StatusCode};
 use hyper_util::rt::TokioIo;
+use std::net::SocketAddr;
 use std::path::Path;
 use tokio::fs::File;
+use tokio::io::AsyncWriteExt;
 use tokio::net::TcpListener;
 use tokio::task::JoinHandle;
 use tokio_util::io::ReaderStream;
 
+/// download downloads file from the given `url` and saves it to file in `file_path`.
+pub async fn download_file(
+    url: &str,
+    local_directory_path: &Path,
+    file: &str,
+    http_peer_list: Vec<(SocketAddr, Option<u16>)>,
+    http_client: &reqwest::Client,
+    file_hash: gevulot_node::types::Hash,
+) -> Result<()> {
+    tracing::trace!("download_file url:{url} local_directory_path:{local_directory_path:?} file:{file} file_hash:{file_hash} http_peer_list:{http_peer_list:?}");
+    let url = reqwest::Url::parse(url)?;
+    let mut resp = match http_client.get(url.clone()).send().await {
+        Ok(resp) => resp,
+        Err(err) => {
+            let peer_urls: Vec<reqwest::Url> = http_peer_list
+                .iter()
+                .filter_map(|(peer, port)| {
+                    port.map(|port| {
+                        //use parse to create an URL, no new method.
+                        let mut url = reqwest::Url::parse("http://localhost").unwrap(); //unwrap always succeed
+                        url.set_ip_host(peer.ip()).unwrap(); //unwrap always succeed
+                        url.set_port(Some(port)).unwrap(); //unwrap always succeed
+                        url.set_path(file); //unwrap always succeed
+                        url
+                    })
+                })
+                .collect();
+            let mut resp = None;
+            for url in peer_urls {
+                if let Ok(val) = http_client.get(url.clone()).send().await {
+                    resp = Some(val);
+                    break;
+                }
+            }
+            match resp {
+                Some(resp) => resp,
+                _ => {
+                    return Err(eyre!(
+                        "Download no host found to download the file: {:?}",
+                        file
+                    ));
+                }
+            }
```

Comment:
```suggestion
            resp.ok_or(Err(eyre!("Download no host found to download the file: {:?}", file)));
```
Could that work? I didn't check the `cargo fmt` or so, but would that make it a bit nicer?